resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -505,32 +508,37 @@ private[yarn] class YarnAllocator(
 
       if (numExecutorsRunning < targetNumExecutors) {
         if (launchContainers) {
-          launcherPool.execute(new Runnable {
-            override def run(): Unit = {
-              try {
-                new ExecutorRunnable(
-                  Some(container),
-                  conf,
-                  sparkConf,
-                  driverUrl,
-                  executorId,
-                  executorHostname,
-                  executorMemory,
-                  executorCores,
-                  appAttemptId.getApplicationId.toString,
-                  securityMgr,
-                  localResources
-                ).run()
-                updateInternalState()
-              } catch {
-                case NonFatal(e) =>
-                  logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
-                  amClient.releaseAssignedContainer(containerId)
+          try {
+            numExecutorToBeLaunched += 1
+            launcherPool.execute(new Runnable {
+              override def run(): Unit = {
+                try {
+                  new ExecutorRunnable(
+                    Some(container),
+                    conf,
+                    sparkConf,
+                    driverUrl,
+                    executorId,
+                    executorHostname,
+                    executorMemory,
+                    executorCores,
+                    appAttemptId.getApplicationId.toString,
+                    securityMgr,
+                    localResources
+                  ).run()
+                  updateInternalState()
+                } catch {
+                  case NonFatal(e) =>
+                    logError(s"Failed to launch executor $executorId on container $containerId", e)
+                    // Assigned container should be released immediately
+                    // to avoid unnecessary resource occupation.
+                    amClient.releaseAssignedContainer(containerId)
+                }
               }
-            }
-          })
+            })
+          } finally {
+            numExecutorToBeLaunched -= 1
```

Comment:
Yes, you're right. When I test the code by experiment, i decrease the `numExecutorToBeLaunched` in the `updateInternalState` function, but I later found this may impact the test.

I will fix this soon.